api/.env

```diff
@@ -0,0 +1,6 @@
+# These environment variables are referneced from within api/docker-compose.yaml to
+# set the user ID and group for files created within the container. A non-empty user
+# and group ID is necessary when running with docker-sync, since the sync tool is actively
+# trying to access files created by the Docker user.
+USER_ID=1001
```

Comment:
Hm... are these right? The previous approach would use your current `$UID` as the user. I don't think 1001 or 82 refers to anything specific I could fine in the code - is this your own UID from your machine?

Reply:
Sorry, I forgot a key bit of documentation here. This file was a workaround to avoid the "permission denied" errors that I was encountering when trying out the earliest version of this branch. Docker-sync needs to use the same UID as the Docker container, which doesn't happen if we leave the 'sync_userid' field blank in docker-sync.yml. This workaround leveraged the ability to have docker-sync and docker-compose share the same .env file (https://docker-sync.readthedocs.io/en/latest/getting-started/configuration.html#environment-variables-support), so they can be guaranteed to use the same user_id this way. The actual numbers were random... I can't remember how I landed on those, I think I saw them commonly in examples in the wild.

This morning I did a bit more digging into docker-sync options, and there is an option sync_userid:'from_host' that causes docker-sync to attach to the current user's userid. I tested this end-to-end and it works – so I'm inclined to switch to that simpler setup, as long as it also works on Linux.

Question for @calbach : is it beneficial / useful / necessary to have docker-compose run as the current user's UID? I see we have some hooks in the code to make this work (I think they trace back to https://github.com/all-of-us/utils/blob/6a6dc7b4bb07c4661b4f6721c1c70cf24ed82042/workbench.rb#L5-L5 where the UID env variable is set, there's also the with-uid.sh script which traces back to https://github.com/all-of-us/workbench/commit/74daf87dd99bea57c568a8a43a892549ce701f67#diff-b3b218df651d0cc4ce869e2a9c36fff5), but I don't know enough historically (or haven't found any docs indicating what strategy we're following) to understand whether or why this might be important.

Reply:
So there's the "user of docker" and the "docker user". Are these different entities?

Reply:
@gjuggler from_host sounds ideal; I can test this out later today. The decision to use the host UID predates me but I can more or less surmise why it was done from some comments and past experience and in my assessment it's a fairly reasonable approach. There are many articles I've seen talking about this issue and no single best practice I've found for handling it.

The issue being mitigated here is that when your docker container creates files within a volume shared with the host filesystem, the default owner of those files will be the docker user. This is a pretty big pain if that's a different UID from your current user. For example, if you run codegen within the container, you'll wind up with a bunch of generated files that the host user cannot remove (without sudo). Often you'll need to run sudo rm -rf ... to get rid of these.

@jaycarlton a docker container defines its own user ID space. This is rarely the same as the host system you're running docker on, as a docker image is defined once and different host systems will have different user spaces. When executing a command within docker, you will specify a user id to act as, or inherit the default for that image.